dask/threaded.py

```diff
@@ -242,40 +241,41 @@ def execute_task(dsk, key, state, queue, results, lock):
     return
 
 
-def finish_task(dsk, key, result, state, results):
+def finish_task(dsk, key, result, state, results, lock):
     """
     Update executation state after a task finishes
 
     Mutates.  This should run atomically (with a lock).
     """
-    state['cache'][key] = result
-    if key in state['ready']:
-        state['ready'].remove(key)
-
-    for dep in state['dependents'][key]:
-        s = state['waiting'][dep]
-        s.remove(key)
-        if not s:
-            del state['waiting'][dep]
-            state['ready'].add(dep)
-
-    for dep in state['dependencies'][key]:
-        if dep in state['waiting_data']:
-            s = state['waiting_data'][dep]
+    with lock:
+        state['cache'][key] = result
+        if key in state['ready']:
+            state['ready'].remove(key)
+
+        for dep in state['dependents'][key]:
+            s = state['waiting'][dep]
             s.remove(key)
-            if not s and dep not in results:
-                if DEBUG:
-                    from chest.core import nbytes
-                    print("Key: %s\tDep: %s\t NBytes: %.2f\t Release" % (key, dep,
-                        sum(map(nbytes, state['cache'].values()) / 1e6)))
-                assert dep in state['cache']
+            if not s:
+                del state['waiting'][dep]
+                state['ready'].add(dep)
+
+        for dep in state['dependencies'][key]:
+            if dep in state['waiting_data']:
+                s = state['waiting_data'][dep]
+                s.remove(key)
+                if not s and dep not in results:
+                    if DEBUG:
+                        from chest.core import nbytes
+                        print("Key: %s\tDep: %s\t NBytes: %.2f\t Release" % (key, dep,
+                            sum(map(nbytes, state['cache'].values()) / 1e6)))
+                    assert dep in state['cache']
+                    release_data(dep, state)
+                    assert dep not in state['cache']
+            elif dep in state['cache'] and dep not in results:
                 release_data(dep, state)
-                assert dep not in state['cache']
-        elif dep in state['cache'] and dep not in results:
-            release_data(dep, state)
 
-    state['finished'].add(key)
-    state['num-active-threads'] -= 1
+        state['finished'].add(key)
+        state['num-active-threads'] -= 1
```

Comment:
I've merged  `state['num-active-threads']` and `jobs` to a single set of running tasks `state['running']` in #8 .  I'm happy to handle the merge conflicts in this case.